### ISSUE TITLE
Enhancement: sanitize Piccolo columns with `secret=True` from PydanticDTO output

### ISSUE THREAD

**Issue by patrickneise** [litestar-org/litestar#2976]
### Summary

Update `generate_field_definitions()` in `PiccoloDTO()` to filter out columns that have `secret=True` set.

Working changes and tests staged [here](https://github.com/patrickneise/litestar/tree/sanitize_secret_columns)

### Basic Example

```python
class AppUser(Table):
    username = Varchar(length=100)
    password = Varchar(length=100, secret=True)
```

PydanticDTO[AppUser] does not contain `password` column.

### Unresolved questions

Motivation came from desire to have sanitized output for `return_dto`, not sure if there are negative impacts for `dto` => not sure if sanitization needs to be conditional based on param passed into PydanticDTO on instantiation (e.g. PydanticDTO(AppUser, exclude_secrets=True)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2976">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2976/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2976/pledge.svg">
</picture>
</a>

**Comment by peterschutt** [comment]
The `dto.Mark.PRIVATE` enum is what we use for this. I'd suggest something like:

```diff
diff --git a/litestar/contrib/piccolo.py b/litestar/contrib/piccolo.py
index 314eb328e..eb495f7c4 100644
--- a/litestar/contrib/piccolo.py
+++ b/litestar/contrib/piccolo.py
@@ -79,10 +79,11 @@ class PiccoloDTO(AbstractDTO[T], Generic[T]):
     @classmethod
     def generate_field_definitions(cls, model_type: type[Table]) -> Generator[DTOFieldDefinition, None, None]:
         for column in model_type._meta.columns:
+            mark = Mark.PRIVATE if column._meta.secret else Mark.READ_ONLY if column._meta.primary_key else None
             yield replace(
                 DTOFieldDefinition.from_field_definition(
                     field_definition=_parse_piccolo_type(column, _create_column_extra(column)),
-                    dto_field=DTOField(mark=Mark.READ_ONLY if column._meta.primary_key else None),
+                    dto_field=DTOField(mark=mark),
                     model_name=model_type.__name__,
                     default_factory=None,
                 ),
```

**Comment by patrickneise** [comment]
Thank you...definitely much cleaner. I have that incorporated and updated tests passing.

Leaves the question of how to handle the difference between `dto` and `return_dto`.

For example:

```python
class Account(Table):
    username = Varchar()
    password = Varchar(secret=True)
```
For routes/controllers:

`return_dto=PiccoloDTO[Account]` behaves as expected, the returned data only contains `username`

`dto=PiccoloDTO[Account]` also only contains `username`.

It seems like that is the expected behavior for fields that are `Mark.PRIVATE`...my question is then, when I want to accept a "private" field in the incoming `dto`, is the correct pattern to create a separate model?

**Comment by peterschutt** [comment]
Perhaps use `Mark.WRITE_ONLY` instead of `Mark.PRIVATE`?

The piccolo DTO implementation doesn't appear to be as featured as the others, as usually there would be a way to declare these things on your model declaration, e.g., with sqlalchemy you can do:

```py
from litestar.dto import dto_field

class Account(Base):
    username: str
    password: str = mapped_column(info=dto_field("write-only"))
```

I don't know enough about piccolo to know if something like that is possible, however there is https://github.com/litestar-org/litestar/pull/2367 which I need to finish that will allow you to use `Annotated[str, DTOField(mark="write_only")]`, and so maybe that will help, but then again, I don't see any examples of piccolo where the columns on the table declaration are annotated.

**Comment by patrickneise** [comment]
Perfect...that might be the way forward until PiccoloDTO gets more robust

```python
class Account[Table]:
    username = Varchar()
    password = Varchar(secret=True)
    
    
    @post("/create", dto=PiccoloDTO[Account], return_dto=PiccoloDTO[Agent])
    aysnc def create_account(data: Account) -> Account
        .....
    ```
    
    Request Body Schema in Swagger docs
    ````
    {
      "username": null,
      "password": null
    }
    ```
    
    Repsonse Body Schema in Swagger docs:
    ```
    {
      "id": null,
      "username": null
    }
    ```
    
    Code is updated in my forked branch if this is good for PR.
    
    Thanks for the help!

**Comment by provinzkraut** [comment]
@patrickneise Even if it fits your use case better, I think as long as we can only implement one behaviour here, marking the fields as `Mark.PRIVATE` makes more sense as a default as it mimics `secret=True` closer in intention and is therefore more likely to fit a wider range of use cases.

@peterschutt?

**Comment by patrickneise** [comment]
@provinzkraut intent is to have this behave similar to piccolo-orm, where "secret" values are only excluded for `select()` queries but still available in the model for input/requests

ideally, PiccoloDTO would implement the same `exclude_secrets` function, but it looks like would require a bit more work at higher level abstraction

the proposal gets part of the way there by only removing the secrets from `response_dto`

**Comment by provinzkraut** [comment]
> @provinzkraut intent is to have this behave similar to piccolo-orm, where "secret" values are only excluded for `select()` queries but still available in the model for input/requests
> 
> ideally, PiccoloDTO would implement the same `exclude_secrets` function, but it looks like would require a bit more work at higher level abstraction
> 
> the proposal gets part of the way there by only removing the secrets from `response_dto`

Thanks for the clarification, that makes sense! Seems I misunderstood the piccolo API there. 
In this case, `Mark.WRITE_ONLY` would be indeed the wat to go!

**Comment by provinzkraut** [comment]
@patrickneise Would you like to create a PR for this? :)

**Comment by provinzkraut** [comment]
Closed in #2976
